inference-engine/src/inference_engine/ie_core.cpp

```diff
@@ -506,18 +426,101 @@ class Core::Impl : public ICore {
+        std::map<std::string, Parameter> getMetricConfig;
+
+        // 0. remove DEVICE_ID key
+        auto deviceIt = compileConfig.find(CONFIG_KEY(DEVICE_ID));
+        if (deviceIt != compileConfig.end()) {
+            getMetricConfig[deviceIt->first] = deviceIt->second;
+            compileConfig.erase(deviceIt);
+        }
+
+        // 1. replace it with DEVICE_ARCHITECTURE value
+        std::vector<std::string> supportedMetricKeys =
+            plugin.GetMetric(METRIC_KEY(SUPPORTED_METRICS), getMetricConfig);
+        auto archIt = std::find(supportedMetricKeys.begin(), supportedMetricKeys.end(),
+            METRIC_KEY(DEVICE_ARCHITECTURE));
+        if (archIt != supportedMetricKeys.end()) {
+            auto value = plugin.GetMetric(METRIC_KEY(DEVICE_ARCHITECTURE), getMetricConfig);
+            compileConfig[METRIC_KEY(DEVICE_ARCHITECTURE)] = value.as<std::string>();
+        } else {
+            // WA: take device name at least if device does not support DEVICE_ARCHITECTURE metric
+            compileConfig[METRIC_KEY(DEVICE_ARCHITECTURE)] = deviceFamily;
+        }
+
+        NetworkCompilationContext context(network, compileConfig);
+        return context.computeHash();
     }
 
     ExecutableNetwork LoadNetwork(const CNNNetwork& network, const std::string& deviceName,
                                   const std::map<std::string, std::string>& config) override {
-        return LoadNetworkImpl(network, deviceName, config, nullptr);
+
+        auto parsed = parseDeviceNameIntoConfig(deviceName, config);
+        auto plugin = GetCPPPluginByName(parsed._deviceName);
+        bool loadedFromCache = false;
+        ExecutableNetwork res;
+        std::string hash;
+        if (DeviceSupportsImportExport(plugin)) {
+            hash = CalculateNetworkHash(network, parsed._deviceName, plugin, config);
+            try {
+                res = LoadNetworkFromCache(hash, plugin, parsed._config, nullptr, loadedFromCache);
+            } catch (const std::exception &) {
+            }
+        }
+
+        if (!loadedFromCache) {
+            res = LoadNetworkImpl(network, plugin, config, nullptr, hash);
+        }
+        return res;
     }
 
-    // TODO: For MULTI plug-in this method can be added to ICore interface
+    // TODO: In future, for MULTI plug-in this method can be added to ICore interface
     ExecutableNetwork LoadNetwork(const std::string& modelPath, const std::string& deviceName,
                                   const std::map<std::string, std::string>& config) {
-        return LoadNetworkImpl(modelPath, deviceName, config, nullptr);
+        auto parsed = parseDeviceNameIntoConfig(deviceName, config);
+        auto plugin = GetCPPPluginByName(parsed._deviceName);
+        bool loadedFromCache = false;
+        ExecutableNetwork res;
+        std::string hash;
+        if (DeviceSupportsImportExport(plugin)) {
+            std::size_t val = std::hash<std::string>()(modelPath + parsed._deviceName);
```

Comment:
why do we cache `parsed._deviceName`? I suppose we need to take into account device architecture instead